src/Spice86.Core/Emulator/CPU/ModRM.cs

```diff
@@ -93,14 +93,24 @@ public class ModRM {
             MemoryAddress = null;
             return;
         }
-        short disp = 0;
-        if (mode == 1) {
-            disp = (sbyte)_cpu.NextUint8();
-        } else if (mode == 2) {
-            disp = (short)_cpu.NextUint16();
-        }
         bool bpForRm6 = mode != 0;
-        MemoryOffset = (ushort)(ComputeOffset(bpForRm6) + disp);
+
+        if (_cpu.AddressSize == 16) {
+            short disp = mode switch {
+                1 => (sbyte)_cpu.NextUint8(),
+                2 => (short)_cpu.NextUint16(),
+                _ => 0
+            };
+            MemoryOffset = (ushort?)(ComputeOffset16(bpForRm6) + disp);
+        } else {
+            int disp = mode switch {
+                1 => (sbyte)_cpu.NextUint8(),
+                2 => (int)_cpu.NextUint32(),
+                _ => 0
+            };
+            MemoryOffset = (ushort?)(ComputeOffset32(mode, _registerMemoryIndex) + disp);
```

Comment:
ComputeOffset32 potentially needs the SIB, so I don't think it can work like that, it needs to be called before disp computation isnt it?

Reply:
Yes you are correct, the displacement bytes are _after_ the SIB byte :(
I think we can fix that by switching the order.